### ISSUE TITLE
iOSScrollBehavior not working in certain cases

### ISSUE THREAD

**Issue by Thorbenandresen** [wix/react-native-keyboard-input#16]
Hi,

I don't have a gist (yet) with a repro but we are experiencing cases where `ScrollToBottomInvertedOnly`and `FixedOffset` iOSScrollBehavior fail. In these cases the keyboard acts like `BehaviorNone. 

* when I open a chat-room while the JS thread is very busy (simulator only)
* when I open a chat-room through resetting my navigation experimental route stack instead of opening it by simply push a new route. 

I know this is kind of a long shot but do you have any idea why this could be happening or can suggest what to investigate?

**Comment by artald** [comment]
There might be an explanation.

The automatic scroll behaviors take place on the native side. To make this happen, your scroll container is identified and a reference to it is kept, so handling the insets/offsets can be done.

As I mentioned before in some issue (I think 😄 ), the asynchronous nature of RN can sometimes cause weird synchronization issues with the UI.
So my guess is that because the bridge is busy and it takes a lot of time for the layout to happen, the reference to the scroll-view is missed and so all the scrolling behavior is lost.

I want you to try something and let me know if you get different results. In Xcode, go to `KeyboardTrackingView.xcodeproj` (under `Libraries`), locate `KeyboardTrackingViewManager.m` and then go to [line 53](https://github.com/wix/react-native-keyboard-tracking-view/blob/master/lib/KeyboardTrackingViewManager.m#L53). Change the number `15` for the `kMaxDeferedInitializeAccessoryViews` const to a higher one, lets say `30`. This controls the number of attempts to "hook" into the native hierarchy.

If this doesn't help we might need to think of a better strategy to do the initial setup so it's always in sync.

**Comment by Thorbenandresen** [comment]
@artald thanks a lot! I tried to set the `kMaxDeferedInitializeAccessoryViews` to 30, 50, 100, 1000 but this did unfortunately not fix the issue.

**Comment by artald** [comment]
Hmm... 🤔 so maybe it's a different issue then, or a different scenario. Thanks for trying.
I wish we had a consistent reproducible example.

**Comment by Thorbenandresen** [comment]
I'll try to cook one up :)

**Comment by artald** [comment]
:)
That would be great, thanks!

**Comment by Thorbenandresen** [comment]
hey @artald, I have finally found a repro: Mount a component (which includes a scroll/list view and the KeyboardAccessoryView), while there still is a another scroll/list view mounted higher up in the view hierarchy.

Checkout my example app with the repro.
https://github.com/Thorbenandresen/keyboard-input-bug-repro

Having another list view mounted on top is common for messaging apps. Take Whats App for example: When you want to start a new chat thread, you will first push an overlay with a list view of your contacts. When you then select a contact from the contact list view, the chat list view is already rendering below while the overlay closes.

**Comment by artald** [comment]
hey @Thorbenandresen, thanks for the effort and the time you took to reproduce it and create a full example! Sorry about the too long explanation (but maybe you'll find it interesting 🤓 )

Without debugging or diving too deep into the code, I think I can tell what's going on.

As I mentioned above, we "hook" into the native hierarchy and find the ScrollView that we then manage to provide all the automatic scroll behaviors and inset/offset handling. We try to get the top scroll container in a breadth first traversal to avoid catching inner scroll views.

You are right that it is a pretty common case to first show a scrolling list of contacts and then push the conversation screen, but in any native app they would be completely different screens with different context (each screen has a separate view-controller that manages the views).

We're using our own [navigation solution](https://github.com/wix/react-native-navigation) and we try to keep it as native as possible. Each native view controller is assigned a separate react root view, so in this case there is no way to get it wrong when finding the scroll container. On the other hand, in a completely JS navigation implementation which I assume you're using - there is only one react root view, so I guess that the other scroll container already exists and we fails to find the correct one.

Bottom line: to solve it we need a way to tell the native code which scroll view is the one we want to manage. At the beginning we actually started with passing a ref to get the scroll but it got things more complicated since you don't always have a direct ref for it, refs can also change and it would sometimes go out of sync. So I will need to play with your demo project and think of a way to do it... thanks!

**Comment by Thorbenandresen** [comment]
@artald thanks so much for the elaborate explanation, indeed interesting! 🤓 

We plan to implement a native navigation solution eventually, but thats big change and might take a while. Implementing a workaround - i.e. unmounting the listviews on the scene right before I mount the chat list view - works well (i.e. not noticeable to the user) if we only have one scene on top. We have one case where we have 3 scenes on top, each of them having a scroll/listview before we go to the chat listview. This is where workarounds turn into ugly hacks 😞 

Do I understand you right, the following solution is not that simple?

```js
<ListView
 ref={(r) => {this.listViewRef = r;}}
...
/>

<KeyboardAccessoryViewComponent
  explicitListViewRef={this.listViewRef} //gets passed over the bridge
  ...
/>

```

**Comment by artald** [comment]
Hey @Thorbenandresen,

Regarding the ref solution, we actually need a ref to the scroll-view and not the list-view, so the ref would need to come from the "scroll component" of the list. But that's not the complexity i was talking about. I was experiencing inconsistent behavior where the ref was still not available at the point when I needed it on the native side. Also, you might not have a direct reference to the scroll-view (for example: the component that handles the list comes from a different package). I would rather find a proper solution.

I have an experimental fix which can possibly correct this issue, but my concern is that it might break other use-cases with other screen layouts.

**If you can, please try this fix manually:**
In `KeyboardTrackingView` -> `KeyboardTrackingViewManager.m` go to the `initializeAccessoryViewsAndHandleInsets`method and replace its implementation with the implementation in [this gist](https://gist.github.com/artald/cf063b71549ecb1d1874ed57f8a83bf8).

Please make sure to remove any workarounds like unmounting the other list-views. Let me know if it solves all cases including the simple case of no other lists (if you have such cases).

Thanks!

**Comment by Thorbenandresen** [comment]
Hi @artald , thanks for the gist!

It fixes the issue on my example app. However, it did not work all in our app - now it is never able to get the correct scrollview (even in cases where it got it before). I will investigate what the difference is.

**Comment by artald** [comment]
Thanks @Thorbenandresen, yeah I suspected that it might not fit all cases.. 😞 

Maybe if I had a similar scenario in the example app I could look into it. In any way, if you investigate it and gain any additional insight - please let me know. Thanks!

**Comment by Thorbenandresen** [comment]
hey @artald , I found the difference. It breaks when I wrap the ListView in an <View>.
I update the example app, see diff here: https://github.com/Thorbenandresen/keyboard-input-bug-repro/commit/1fce4965219f3ac2776a0bf25948bd57ae32dcff

**Comment by artald** [comment]
Yes, it makes sense it would break in this case (though I didn't actually checked it...).

The "experimental" fix was to check that the keyboard accessory view and the scroll-view have the same parent. So if you just wrap the list view (which contains the scroll view) with another view - it breaks for all cases. That's what I meant that I had concerns that this fix might break screens with different layouts...

**Comment by Thorbenandresen** [comment]
@artald - oh that is what the fix does! 👍  I think I can refactor my code so that scroll-view and the keyboard accessory have the same parent. 

Is it possible to ship these changes as non-defaults? E.g. by enabling this behavior via a prop?

**Comment by artald** [comment]
@Thorbenandresen yup, that's what it does... but please don't refactor your code just yet, I don't want you to do all this work for nothing :)

I want to first see if there's no better way to solve it. If there's no other solution I guess this fix could be enabled via a prop which will be an opt-in.

Do you think that after refactoring it would fix all cases for you?

**Comment by Thorbenandresen** [comment]
@artald - I only had to make minor changes which I gonna keep regardless :) It looks like everything is working just fine 👍

**Comment by artald** [comment]
@Thorbenandresen that's great. I don't have a better solution yet, so we'll go with this fix for now.
I will try to it sometime this week. I hope it's not too urgent..

**Comment by Thorbenandresen** [comment]
@artald thanks and no problem, I can use my own fork until then.

**Comment by Thorbenandresen** [comment]
@artald I just upgraded from 3.0.10 to 3.0.20 and I currently have same issue with the content insets/offsets when both the scroll-view and the keyboard accessory share the same parent. When I open the keyboard, the scrollview lags behind for little bit, i.e. there is no perfect fixed offset. If they do not share the same parent (or if I use 3.0.10), the fixed offset works fine. 

I have a couple of other ideas I can try, but if you happen to know why that is the case and how to fix it, please let me know :)

**Comment by artald** [comment]
@Thorbenandresen the 2 packages are indeed coupled. `keyboard-tracking-view` is the one who's actually providing the scroll behavior, manages the scroll-view insets/offsets etc. The `react-native-keyboard-input` is all about providing the custom keyboard functionality and gluing it all together in an easy to use manner.

So the relevant changes are the ones made to `react-native-keyboard-tracking-view`. Between version `3.0.6` and the latest `3.0.11` we had the following changes:

- `3.0.7`: support RN versions > 40. This is just updating native imports, not supposed to hurt anything
- `3.0.8`: fix for issue #13 that you opened to stop current scroll before scrolling to bottom
- `3.0.9`:  fix for issue #15 that you opened with interactive reveal.
- `3.0.10`: **(This is the interesting one)** fixed scrolling behavior for various edge cases.
- `3.0.11`: support disabling the automatically managed scroll via a prop

If something got accidentally broken, it is in version `3.0.10` of `react-native-keyboard-tracking-view`. We noticed some very weird cases of insets/offsets not working as expected, for example when there is very little content in the scroll-view/list-view, and also wrong offset being applied on various cases when the screen which contains the scroll appears for the first time. We took a different approach for applying the insets/offsets logic, so I hope it didn't break anything.

That said, **I couldn't reproduce the issue** you've mentioned. It's probably best to **open a new issue** for this as you suggested. As always - if you can update the existing demo project (keyboard-input-bug-repro) with a repro it would be greatly appreciated :)

**Comment by artald** [comment]
@Thorbenandresen I just published `react-native-keyboard-input@3.0.21` along with `react-native-keyboard-tracking-view@3.0.12`.

You can now use the new prop `requiresSameParentToManageScrollView` on `KeyboardAccessoryView`. This should solve the problem with multiple containers shown in the same react root view as demonstrated in https://github.com/Thorbenandresen/keyboard-input-bug-repro by requiring that both the scroll-view container and the content of `KeyboardAccessoryView` have the same parent (same containing view). This will allow it to identify the exact scroll container.

Please verify that this fixes this issue for you, thanks!

**Comment by Thorbenandresen** [comment]
Hey @artald,

thanks a lot! I have hard time pinpointing what the reason is, but I currently don't have this issue any longer with the following setup

* RN 43
* keyboard-input v3.0.21.
* keyboard-tracking-view v3.0.9 (added as first level dep, to work around https://github.com/wix/react-native-keyboard-input/issues/22)
* ScrollView and KeyboardAccessoryView sharing the same parent

This means that the keyboard is working without the latest addition to the keyboard-tracking-view (i.e. requiresSameParentToManageScrollView prop). Strange 🤔
